dspace-api/src/main/java/org/dspace/discovery/indexobject/IndexFactoryImpl.java

```diff
@@ -74,32 +79,59 @@ public void writeDocument(Context context, T indexableObject, SolrInputDocument
      * Write the document to the index under the appropriate unique identifier.
      *
      * @param doc     the solr document to be written to the server
-     * @param streams list of bitstream content streams    DiscoverQueryBuilderTest.java:285
+     * @param streams list of bitstream content streams
      * @throws IOException A general class of exceptions produced by failed or interrupted I/O operations.
      */
     protected void writeDocument(SolrInputDocument doc, FullTextContentStreams streams)
             throws IOException, SolrServerException {
         final SolrClient solr = solrSearchCore.getSolr();
         if (solr != null) {
             if (streams != null && !streams.isEmpty()) {
-                ContentStreamUpdateRequest req = new ContentStreamUpdateRequest("/update/extract");
-                req.addContentStream(streams);
-
-                ModifiableSolrParams params = new ModifiableSolrParams();
+                // limit full text indexing to first 100,000 characters unless configured otherwise
+                final int charLimit = DSpaceServicesFactory.getInstance().getConfigurationService()
+                                                           .getIntProperty("discovery.solr.fulltext.charLimit",
+                                                                           100000);
+
+                // Use Tika's Text parser as the streams are always from the TEXT bundle (i.e. already extracted text)
+                // TODO: We may wish to consider using Tika to extract the text in the future.
+                TextAndCSVParser tikaParser = new TextAndCSVParser();
+                BodyContentHandler tikaHandler = new BodyContentHandler(charLimit);
+                Metadata tikaMetadata = new Metadata();
+                ParseContext tikaContext = new ParseContext();
+
+                // Use Apache Tika to parse the full text stream
+                try {
+                    tikaParser.parse(streams.getStream(), tikaHandler, tikaMetadata, tikaContext);
+                } catch (SAXException saxe) {
+                    // Check if this SAXException is just a notice that this file was longer than the character limit.
+                    // Unfortunately there is not a unique, public exception type to catch here. This error is thrown
+                    // by Tika's WriteOutContentHandler when it encounters a document longer than the char limit
+                    // https://github.com/apache/tika/blob/main/tika-core/src/main/java/org/apache/tika/sax/WriteOutContentHandler.java
+                    if (saxe.getMessage().contains("limit has been reached")) {
+                        // log that we only indexed up to that configured limit
+                        log.info("Full text is larger than the configured limit (discovery.solr.fulltext.charLimit)."
+                                     + " Only the first " + charLimit + " characters were indexed.");
```

Comment:
Log4j2's parameter substitution would reduce this expensive multiple-string-concatenation expression to a compile-time constant.